Python/ogusa/tax.py

```diff
@@ -250,40 +289,80 @@ def MTR_labor(r, b, w, e, n, factor, mtrx_params):
         tau = derivative of tau_income w.r.t. labor income (various length array or scalar)
     '''
 
-    if mtrx_params.ndim == 2: 
-        A = mtrx_params[:,0]
-        B = mtrx_params[:,1]
-        C = mtrx_params[:,2]
-        D = mtrx_params[:,3]
-        E = mtrx_params[:,4]
-        F = mtrx_params[:,5]
-        max_x = mtrx_params[:,6]
-        min_x = mtrx_params[:,7]
-        max_y = mtrx_params[:,8]
-        min_y = mtrx_params[:,9]
-    if mtrx_params.ndim == 1: 
-        A = mtrx_params[0]
-        B = mtrx_params[1]
-        C = mtrx_params[2]
-        D = mtrx_params[3]
-        E = mtrx_params[4]
-        F = mtrx_params[5]
-        max_x = mtrx_params[6]
-        min_x = mtrx_params[7]
-        max_y = mtrx_params[8]
-        min_y = mtrx_params[9]
-  
-    x = (w*e*n)*factor
-    y = (r*b)*factor
-    I = x+y
-
-    phi = x/I
-    Phi = phi*(max_x-min_x) + (1-phi)*(max_y-min_y)
-    K = phi*min_x + (1-phi)*min_y
-
-    num = (A*(x**2)) + (B*(y**2)) + (C*x*y) + (D*x) + (E*y)
-    denom = (A*(x**2)) + (B*(y**2)) + (C*x*y) + (D*x) + (E*y) + F
-    tau =  (Phi*(num/denom)) + K
+    if analytical_mtrs:
+        if etr_params.ndim == 2: 
+            A = etr_params[:,0]
+            B = etr_params[:,1]
+            C = etr_params[:,2]
+            D = etr_params[:,3]
+            E = etr_params[:,4]
+            F = etr_params[:,5]
+            max_x = etr_params[:,6]
+            min_x = etr_params[:,7]
+            max_y = etr_params[:,8]
+            min_y = etr_params[:,9]
+        if etr_params.ndim == 1: 
+            A = etr_params[0]
+            B = etr_params[1]
+            C = etr_params[2]
+            D = etr_params[3]
+            E = etr_params[4]
+            F = etr_params[5]
+            max_x = etr_params[6]
+            min_x = etr_params[7]
+            max_y = etr_params[8]
+            min_y = etr_params[9]
+        
+        x = (w*e*n)*factor
+        y = (r*b)*factor
+        I = x+y
+
+        num = (A*(x**2)) + (B*(y**2)) + (C*x*y) + (D*x) + (E*y)
+        denom = (A*(x**2)) + (B*(y**2)) + (C*x*y) + (D*x) + (E*y) + F
+        Lambda = num/denom
+
+        d_num = (2*A*x + C*y + D)*F
+        d_denom = ((A*(x**2)) + (B*(y**2)) + (C*x*y) + (D*x) + (E*y) + F)**2
+        d_Lambda = d_num/d_denom
+
+        tau =  (max_x-min_x)*Lambda + (x*(max_x-min_x) + y*(max_y-min_y))*d_Lambda + min_x
+
+    if analytical_mtrs:
```

Comment:
is this right? At line 292 you have `if analytical_mtrs`, but now here again you have `if analytical_mtrs`?